src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs

```diff
@@ -298,8 +315,28 @@ private async Task ProcessOperationRequestAsync(RequestMessage request)
                 this.context.CurrentOperationResponse.SetExceptionResult(
                     e,
                     this.context.CurrentOperation.Operation,
-                    this.EntityId,
                     this.dataConverter);
+
+                operationFailed = true;
+            }
+
+            if (this.RollbackFailedOperations)
+            {
+                // we write back the entity state after each successful operation
+                if (!operationFailed)
+                {
+                    if (!this.context.TryWriteback(out var errorResponseMessage))
+                    {
+                        // state serialization failed; create error response and roll back.
+                        this.context.CurrentOperationResponse = errorResponseMessage;
+                        operationFailed = true;
+                    }
+                }
+                else
```

Comment:
This is not quite right I think, we want the rollback to happen if the serialization fails. I should also add a test that catches this.